worlds/messenger/__init__.py

```diff
@@ -116,44 +210,132 @@ def create_items(self) -> None:
             total_seals = min(len(self.multiworld.get_unfilled_locations(self.player)) - len(itempool),
                               self.options.total_seals.value)
             if total_seals < self.total_seals:
-                logging.warning(f"Not enough locations for total seals setting "
-                                f"({self.options.total_seals}). Adjusting to {total_seals}")
+                logging.warning(
+                    f"Not enough locations for total seals setting "
+                    f"({self.options.total_seals}). Adjusting to {total_seals}"
+                )
                 self.total_seals = total_seals
             self.required_seals = int(self.options.percent_seals_required.value / 100 * self.total_seals)
 
             seals = [self.create_item("Power Seal") for _ in range(self.total_seals)]
-            for i in range(self.required_seals):
-                seals[i].classification = ItemClassification.progression_skip_balancing
             itempool += seals
 
+        self.multiworld.itempool += itempool
         remaining_fill = len(self.multiworld.get_unfilled_locations(self.player)) - len(itempool)
         if remaining_fill < 10:
             self._filler_items = self.random.choices(
-                                      list(FILLER)[2:],
-                                      weights=list(FILLER.values())[2:],
-                                      k=remaining_fill
+                list(FILLER)[2:],
+                weights=list(FILLER.values())[2:],
+                k=remaining_fill
             )
-        itempool += [self.create_filler() for _ in range(remaining_fill)]
+        filler = [self.create_filler() for _ in range(remaining_fill)]
 
-        self.multiworld.itempool += itempool
+        self.multiworld.itempool += filler
 
     def set_rules(self) -> None:
         logic = self.options.logic_level
         if logic == Logic.option_normal:
             MessengerRules(self).set_messenger_rules()
-        elif logic == Logic.option_hard:
-            MessengerHardRules(self).set_messenger_rules()
         else:
-            MessengerOOBRules(self).set_messenger_rules()
+            MessengerHardRules(self).set_messenger_rules()
+        # else:
+        #     MessengerOOBRules(self).set_messenger_rules()
+
+        add_closed_portal_reqs(self)
+        # i need portal shuffle to happen after rules exist so i can validate it
+        attempts = 5
+        if self.options.shuffle_portals:
+            self.portal_mapping = []
+            self.spoiler_portal_mapping = {}
+            for _ in range(attempts):
+                disconnect_portals(self)
+                shuffle_portals(self)
+                if validate_portals(self):
+                    break
+            # failsafe mostly for invalid plandoed portals with no transition shuffle
+            else:
+                raise RuntimeError("Unable to generate valid portal output.")
```

Comment:
shuffled portals generally open the game up more than the vanilla layout does. the validation step will always pass if transitions are shuffled (though that's disabled in this current PR), and if transitions aren't shuffled it always succeeded on the first try in my testing. it's still *possible* that since it does the selection completely randomly that it might come up with an invalid selection, but it's still very unlikely. the failsafe is mostly just here in case someone plandoes 5-6 of the available 6 portals and none of them give any sphere 1 access.